File: packages/contracts/src/chatBackendsContract.mjs
import chatBackendsJson from "./chat-backends.json" with { type: "json" };

export const CHAT_BACKENDS_CONTRACT = deepFreeze(chatBackendsJson);

export const CHAT_BACKENDS = CHAT_BACKENDS_CONTRACT.chatBackends;
export const DEFAULT_CHAT_BACKEND = CHAT_BACKENDS_CONTRACT.defaultBackend;
export const CHAT_BACKEND_LABELS = CHAT_BACKENDS_CONTRACT.backendLabels;
export const DEFAULT_MODEL_BY_BACKEND = CHAT_BACKENDS_CONTRACT.defaultModels;
export const MODEL_OPTIONS_BY_BACKEND = CHAT_BACKENDS_CONTRACT.backendModels;
export const ALLOWED_MODEL_PREFIXES_BY_BACKEND =
  CHAT_BACKENDS_CONTRACT.allowedModelPrefixes;
export const REASONING_EFFORTS = CHAT_BACKENDS_CONTRACT.reasoningEfforts;
export const BACKEND_REASONING_EFFORTS =
  CHAT_BACKENDS_CONTRACT.backendReasoningEfforts;
export const CODEX_REASONING_EFFORTS = BACKEND_REASONING_EFFORTS.codex;
export const DIRECT_DEFAULT_URLS = CHAT_BACKENDS_CONTRACT.directUrls;
export const API_KEY_ENV_BY_BACKEND = CHAT_BACKENDS_CONTRACT.apiKeyEnv;
export const PROVIDER_STORE_KEY_BY_BACKEND =
  CHAT_BACKENDS_CONTRACT.providerStoreKeys;
export const ROUTER_STORE_KEY_BY_BACKEND =
  CHAT_BACKENDS_CONTRACT.routerStoreKeys;
export const ROUTER_MODES_BY_BACKEND =
  CHAT_BACKENDS_CONTRACT.backendRouterModes;
export const DEFAULT_ROUTER_MODE_BY_BACKEND =
  CHAT_BACKENDS_CONTRACT.defaultRouterModes;
export const ROUTER_MODE_LABELS = CHAT_BACKENDS_CONTRACT.routerModeLabels;
export const ROUTER_MODES_USING_API_CONFIG =
  CHAT_BACKENDS_CONTRACT.routerModesUsingApiConfig;
export const ROUTER_MODES_USING_CODEX_ACCOUNT =
  CHAT_BACKENDS_CONTRACT.routerModesUsingCodexAccount;
export const API_DESCRIPTION_BY_BACKEND =
  CHAT_BACKENDS_CONTRACT.apiDescriptions;
export const CONNECTION_MODES = CHAT_BACKENDS_CONTRACT.connectionModes;
export const CONNECTION_MODES_USING_API_KEY =
  CHAT_BACKENDS_CONTRACT.connectionModesUsingApiKey;
export const CONNECTION_MODES_USING_DEFAULT_API =
  CHAT_BACKENDS_CONTRACT.connectionModesUsingDefaultApi;
export const CONNECTION_MODES_USING_CUSTOM_URL =
  CHAT_BACKENDS_CONTRACT.connectionModesUsingCustomUrl;
export const CONNECTION_MODES_USING_CODEX_ACCOUNT =
  CHAT_BACKENDS_CONTRACT.connectionModesUsingCodexAccount;
export const UNCONFIGURED_CONNECTION_MODES =
  CHAT_BACKENDS_CONTRACT.unconfiguredConnectionModes;

function deepFreeze(value) {
  if (!value || typeof value !== "object") return value;
  for (const child of Object.values(value)) {
    deepFreeze(child);
  }
  return Object.freeze(value);
}

File: apps/desktop/src-tauri/src/chat_backends_contract.rs
use std::collections::BTreeMap;
use std::sync::OnceLock;

use serde::Deserialize;

const CHAT_BACKENDS_JSON: &str =
    include_str!("../../../../packages/contracts/src/chat-backends.json");

static CHAT_BACKENDS_CONTRACT: OnceLock<ChatBackendsContract> = OnceLock::new();

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ChatBackendsContract {
    pub(crate) chat_backends: Vec<String>,
    pub(crate) default_backend: String,
    pub(crate) default_models: BTreeMap<String, String>,
    pub(crate) backend_models: BTreeMap<String, Vec<ChatModelOptionContract>>,
    pub(crate) allowed_model_prefixes: BTreeMap<String, Vec<String>>,
    pub(crate) reasoning_efforts: Vec<String>,
    pub(crate) backend_reasoning_efforts: BTreeMap<String, Vec<String>>,
    pub(crate) direct_urls: BTreeMap<String, String>,
    pub(crate) api_key_env: BTreeMap<String, String>,
    pub(crate) provider_store_keys: BTreeMap<String, String>,
    pub(crate) router_store_keys: BTreeMap<String, String>,
    pub(crate) backend_router_modes: BTreeMap<String, Vec<String>>,
    pub(crate) default_router_modes: BTreeMap<String, String>,
    pub(crate) connection_modes_using_api_key: Vec<String>,
    pub(crate) connection_modes_using_default_api: Vec<String>,
    pub(crate) connection_modes_using_custom_url: Vec<String>,
    pub(crate) connection_modes_using_codex_account: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ChatModelOptionContract {
    pub(crate) id: String,
    pub(crate) label: String,
}

pub(crate) fn chat_backends_contract() -> &'static ChatBackendsContract {
    CHAT_BACKENDS_CONTRACT.get_or_init(|| {
        crate::contract_manifest::parse_contract_json(CHAT_BACKENDS_JSON, "chat-backends.json")
    })
}
